// @suprsend/react-inbox/dist/index.js
function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var React = require('react');
var React__default = _interopDefault(React);
var SuprsendJSInbox = _interopDefault(require('@suprsend/js-inbox'));
var reactPopper = require('react-popper');
var styled = _interopDefault(require('@emotion/styled'));
var react = require('@emotion/react');
var Markdown = _interopDefault(require('react-markdown'));
var TimeAgo = _interopDefault(require('react-timeago'));
var InfiniteScroll = _interopDefault(require('react-infinite-scroll-component'));
var reactToastify = require('react-toastify');

function _extends() {
  _extends = Object.assign ? Object.assign.bind() : function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];
      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }
    return target;
  };
  return _extends.apply(this, arguments);
}
function _taggedTemplateLiteralLoose(strings, raw) {
  if (!raw) {
    raw = strings.slice(0);
  }
  strings.raw = raw;
  return strings;
}

var _templateObject, _templateObject2, _templateObject3;
var lightColors = {
  primary: '#2E70E8',
  primaryText: '#1E293B',
  secondaryText: '#475569',
  border: '#E2E8F0',
  main: '#FFFFFF',
  error: '#B42318'
};
var darkColors = {
  primary: '#2E70E8',
  primaryText: '#EFEFEF',
  secondaryText: '#CBD5E1',
  border: '#3A4A61',
  main: '#1D2635',
  error: '#F97066'
};
var CText = styled.p(_templateObject || (_templateObject = _taggedTemplateLiteralLoose(["\n  font-size: 14px;\n  font-weight: 400;\n  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Helvetica,\n    Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol';\n  line-height: 20px;\n  color: ", ";\n  margin: 0px;\n"])), lightColors.primaryText);
var HelperText = styled(CText)(_templateObject2 || (_templateObject2 = _taggedTemplateLiteralLoose(["\n  font-size: 12px;\n  color: ", ";\n"])), lightColors.secondaryText);
var HeadingText = styled(CText)(_templateObject3 || (_templateObject3 = _taggedTemplateLiteralLoose(["\n  font-size: 16px;\n  font-weight: 500;\n"])));
var darkTheme = {
  bell: {
    color: '#fff'
  },
  badge: {
    backgroundColor: darkColors.primary
  },
  header: {
    container: {
      backgroundColor: darkColors.main,
      borderBottom: "0.5px solid " + darkColors.border,
      boxShadow: '0 0 5px 0 rgba(0, 0, 0, 0.5)'
    },
    headerText: {
      color: darkColors.primaryText
    },
    markAllReadText: {
      color: darkColors.primary
    }
  },
  tabs: {
    color: darkColors.primaryText,
    unselectedColor: darkColors.secondaryText + 'D9',
    bottomColor: darkColors.primary,
    badgeColor: 'rgba(100, 116, 139, 0.5)',
    badgeText: darkColors.primaryText
  },
  notificationsContainer: {
    container: {
      backgroundColor: darkColors.main,
      borderColor: darkColors.border
    },
    noNotificationsText: {
      color: darkColors.primaryText
    },
    noNotificationsSubtext: {
      color: darkColors.secondaryText
    },
    loader: {
      color: darkColors.primary
    }
  },
  notification: {
    container: {
      borderBottom: "1px solid " + darkColors.border,
      readBackgroundColor: darkColors.main,
      unreadBackgroundColor: '#273244',
      hoverBackgroundColor: '#2D3A4D'
    },
    pinnedText: {
      color: darkColors === null || darkColors === void 0 ? void 0 : darkColors.secondaryText
    },
    headerText: {
      color: darkColors.primaryText
    },
    bodyText: {
      color: darkColors.secondaryText,
      blockquoteColor: 'rgba(100, 116, 139, 0.5)'
    },
    unseenDot: {
      backgroundColor: darkColors.primary
    },
    createdOnText: {
      color: darkColors.secondaryText
    },
    subtext: {
      color: '#94a3b8'
    },
    actions: [{
      container: {
        backgroundColor: darkColors.primary
      }
    }, {
      container: {
        borderColor: darkColors.border,
        backgroundColor: 'transparent',
        hoverBackgroundColor: darkColors.main
      },
      text: {
        color: darkColors.secondaryText
      }
    }],
    expiresText: {
      backgroundColor: 'rgba(100, 116, 139, 0.5)',
      color: darkColors.secondaryText,
      expiringBackgroundColor: 'rgba(217, 45, 32, 0.15)',
      expiringColor: darkColors.error
    },
    actionsMenuIcon: {
      color: darkColors.secondaryText,
      hoverBackgroundColor: 'rgba(100, 116, 139, 0.5)'
    },
    actionsMenu: {
      backgroundColor: darkColors.main,
      borderColor: darkColors.border
    },
    actionsMenuItem: {
      hoverBackgroundColor: 'rgba(100, 116, 139, 0.2)'
    },
    actionsMenuItemIcon: {
      color: darkColors.secondaryText
    },
    actionsMenuItemText: {
      color: darkColors.secondaryText
    }
  },
  toast: {
    container: {
      backgroundColor: darkColors.main,
      borderColor: darkColors.border
    },
    headerText: {
      color: darkColors.primaryText
    },
    bodyText: {
      color: darkColors.secondaryText,
      blockquoteColor: darkColors.border
    }
  }
};

var DefaultBellIcon = function DefaultBellIcon(_ref) {
  var style = _ref.style;
  var height = (style === null || style === void 0 ? void 0 : style.height) || 24;
  var width = (style === null || style === void 0 ? void 0 : style.width) || 24;
  var color = (style === null || style === void 0 ? void 0 : style.color) || lightColors.primaryText;
  if ((style === null || style === void 0 ? void 0 : style.bellType) === 'solid') {
    return /*#__PURE__*/React__default.createElement("svg", {
      xmlns: "http://www.w3.org/2000/svg",
      width: width,
      height: height,
      color: color,
      viewBox: "0 0 20 20",
      fill: "currentColor"
    }, /*#__PURE__*/React__default.createElement("path", {
      d: "M10 2a6 6 0 00-6 6v3.586l-.707.707A1 1 0 004 14h12a1 1 0 00.707-1.707L16 11.586V8a6 6 0 00-6-6zM10 18a3 3 0 01-3-3h6a3 3 0 01-3 3z"
    }));
  }
  return /*#__PURE__*/React__default.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    width: width,
    height: height,
    color: color,
    fill: "none",
    viewBox: "0 0 24 24",
    stroke: "currentColor",
    strokeWidth: "2"
  }, /*#__PURE__*/React__default.createElement("path", {
    strokeLinecap: "round",
    strokeLinejoin: "round",
    d: "M15 17h5l-1.405-1.405A2.032 2.032 0 0118 14.158V11a6.002 6.002 0 00-4-5.659V5a2 2 0 10-4 0v.341C7.67 6.165 6 8.388 6 11v3.159c0 .538-.214 1.055-.595 1.436L4 17h5m6 0v1a3 3 0 11-6 0v-1m6 0H9"
  }));
};

var InboxContext = React.createContext({});
var InboxThemeContext = React.createContext({});
function useInbox() {
  return React.useContext(InboxContext);
}
function useTheme() {
  return React.useContext(InboxThemeContext);
}

function Bell() {
  var _useInbox = useInbox(),
    bellComponent = _useInbox.bellComponent;
  var _useTheme = useTheme(),
    bell = _useTheme.bell;
  if (bellComponent) {
    var BellComponent = bellComponent;
    return /*#__PURE__*/React__default.createElement(BellComponent, null);
  }
  return /*#__PURE__*/React__default.createElement(DefaultBellIcon, {
    style: bell
  });
}

var _templateObject$1;
function Badge() {
  var _useInbox = useInbox(),
    count = _useInbox.notificationsData.count,
    badgeComponent = _useInbox.badgeComponent;
  var _useTheme = useTheme(),
    badge = _useTheme.badge;
  if (count > 0) {
    if (badgeComponent) {
      var BagdeComponent = badgeComponent;
      return /*#__PURE__*/React__default.createElement(BagdeComponent, {
        count: count
      });
    }
    return /*#__PURE__*/React__default.createElement(CountText, {
      style: badge
    }, count);
  }
  return null;
}
var CountText = styled(CText)(_templateObject$1 || (_templateObject$1 = _taggedTemplateLiteralLoose(["\n  position: absolute;\n  right: -3px;\n  top: -7px;\n  display: inline-block;\n  font-size: 10px;\n  line-height: 1;\n  padding: 3px 6px;\n  border-radius: 50%;\n  background-color: ", ";\n  color: ", ";\n  text-align: center;\n"])), lightColors.primary, lightColors.main);

var isImgUrl = function isImgUrl(url) {
  try {
    if (url) {
      var img = new window.Image();
      img.src = url;
      return Promise.resolve(new Promise(function (resolve) {
        img.onerror = function () {
          return resolve(false);
        };
        img.onload = function () {
          return resolve(true);
        };
      }));
    }
    return Promise.resolve();
  } catch (e) {
    return Promise.reject(e);
  }
};
function isObject(item) {
  return item && typeof item === 'object' && !Array.isArray(item);
}
function mergeDeep(target, source) {
  var output = Object.assign({}, target);
  if (isObject(target) && isObject(source)) {
    Object.keys(source).forEach(function (key) {
      if (isObject(source[key])) {
        var _Object$assign;
        if (!(key in target)) Object.assign(output, (_Object$assign = {}, _Object$assign[key] = source[key], _Object$assign));else output[key] = mergeDeep(target[key], source[key]);
      } else {
        var _Object$assign2;
        Object.assign(output, (_Object$assign2 = {}, _Object$assign2[key] = source[key], _Object$assign2));
      }
    });
  }
  return output;
}
function formatActionLink(link) {
  if (!link) return;
  if (link.startsWith('http') || link.startsWith('/')) {
    return link;
  } else {
    return "https://" + link;
  }
}
function getLongFormattedTime(value, unit) {
  switch (unit) {
    case 'second':
      return 'a minute';
    case 'minute':
      return value === 1 ? value + " minute" : value + " minutes";
    case 'hour':
      return value === 1 ? value + " hour" : value + " hours";
    case 'day':
      return value === 1 ? value + " day" : value + " days";
    case 'week':
      return value === 1 ? value + " week" : value + " weeks";
    case 'month':
      return value === 1 ? value + " month" : value + " months";
    case 'year':
      return value === 1 ? value + " year" : value + " years";
    default:
      return value;
  }
}
function getShortFormattedTime(value, unit) {
  switch (unit) {
    case 'second':
      return '1m';
    case 'minute':
      return value + "m";
    case 'hour':
      return value + "h";
    case 'day':
      return value + "d";
    case 'week':
      return value + "w";
    case 'month':
      return value + "mo";
    case 'year':
      return value + "y";
    default:
      return value;
  }
}

var AvatarIcon = function AvatarIcon(_ref) {
  var type = _ref.type;
  if (type === 'dark') {
    return /*#__PURE__*/React__default.createElement("svg", {
      width: "36",
      height: "36",
      viewBox: "0 0 36 36",
      fill: "none",
      xmlns: "http://www.w3.org/2000/svg"
    }, /*#__PURE__*/React__default.createElement("circle", {
      cx: "18",
      cy: "18",
      r: "18",
      fill: "#334155"
    }), /*#__PURE__*/React__default.createElement("path", {
      d: "M23.8641 14.2836C23.8641 17.3522 21.4556 19.8399 18.4844 19.8399C15.5132 19.8399 13.1047 17.3522 13.1047 14.2836C13.1047 11.215 15.5132 8.72754 18.4844 8.72754C21.4556 8.72754 23.8641 11.215 23.8641 14.2836Z",
      fill: "#94A3B8"
    }), /*#__PURE__*/React__default.createElement("path", {
      d: "M23.8643 20.9316H13.1051C11.1039 20.9316 9.37694 22.0734 8.4502 23.75C10.6277 27.0754 14.3128 29.2659 18.4848 29.2659C22.6568 29.2659 26.3418 27.0754 28.5195 23.75C27.5913 22.0734 25.866 20.9316 23.8645 20.9316H23.8643Z",
      fill: "#94A3B8"
    }));
  }
  return /*#__PURE__*/React__default.createElement("svg", {
    width: "36",
    height: "36",
    viewBox: "0 0 36 36",
    fill: "none",
    xmlns: "http://www.w3.org/2000/svg"
  }, /*#__PURE__*/React__default.createElement("circle", {
    cx: "18",
    cy: "18",
    r: "18",
    fill: "#CBD5E1"
  }), /*#__PURE__*/React__default.createElement("path", {
    d: "M23.8641 14.2836C23.8641 17.3522 21.4556 19.8399 18.4844 19.8399C15.5132 19.8399 13.1047 17.3522 13.1047 14.2836C13.1047 11.215 15.5132 8.72754 18.4844 8.72754C21.4556 8.72754 23.8641 11.215 23.8641 14.2836Z",
    fill: "white"
  }), /*#__PURE__*/React__default.createElement("path", {
    d: "M23.8643 20.9316H13.1051C11.1039 20.9316 9.37694 22.0734 8.4502 23.75C10.6277 27.0754 14.3128 29.2659 18.4848 29.2659C22.6568 29.2659 26.3418 27.0754 28.5195 23.75C27.5913 22.0734 25.866 20.9316 23.8645 20.9316H23.8643Z",
    fill: "white"
  }));
};

var MoreIcon = function MoreIcon(_ref) {
  var style = _ref.style;
  var height = (style === null || style === void 0 ? void 0 : style.height) || 20;
  var width = (style === null || style === void 0 ? void 0 : style.width) || 20;
  var color = (style === null || style === void 0 ? void 0 : style.color) || lightColors.secondaryText;
  return /*#__PURE__*/React__default.createElement("svg", {
    width: width,
    height: height,
    viewBox: "0 0 24 24",
    fill: "none",
    xmlns: "http://www.w3.org/2000/svg"
  }, /*#__PURE__*/React__default.createElement("path", {
    d: "M12 13C12.5523 13 13 12.5523 13 12C13 11.4477 12.5523 11 12 11C11.4477 11 11 11.4477 11 12C11 12.5523 11.4477 13 12 13Z",
    stroke: color,
    strokeWidth: "2",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React__default.createElement("path", {
    d: "M19 13C19.5523 13 20 12.5523 20 12C20 11.4477 19.5523 11 19 11C18.4477 11 18 11.4477 18 12C18 12.5523 18.4477 13 19 13Z",
    stroke: color,
    strokeWidth: "2",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React__default.createElement("path", {
    d: "M5 13C5.55228 13 6 12.5523 6 12C6 11.4477 5.55228 11 5 11C4.44772 11 4 11.4477 4 12C4 12.5523 4.44772 13 5 13Z",
    stroke: color,
    strokeWidth: "2",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }));
};

var PinnedNotificationIcon = function PinnedNotificationIcon(_ref) {
  var style = _ref.style;
  var height = (style === null || style === void 0 ? void 0 : style.height) || 13;
  var width = (style === null || style === void 0 ? void 0 : style.width) || 13;
  var color = (style === null || style === void 0 ? void 0 : style.color) || '#DC6803';
  return /*#__PURE__*/React__default.createElement("svg", {
    width: width,
    height: height,
    viewBox: "0 0 16 17",
    fill: "none",
    xmlns: "http://www.w3.org/2000/svg"
  }, /*#__PURE__*/React__default.createElement("path", {
    d: "M7.58984 2.9305C7.38193 2.44536 7.27797 2.2028 7.10998 2.09261C6.96308 1.99626 6.78408 1.96179 6.6119 1.99669C6.415 2.03661 6.22839 2.22321 5.85518 2.59643L2.42642 6.02519C2.0532 6.39841 1.86659 6.58502 1.82668 6.78191C1.79177 6.95409 1.82624 7.13309 1.9226 7.28C2.03278 7.44799 2.27535 7.55194 2.76048 7.75986L4.42339 8.47253C4.49393 8.50277 4.5292 8.51788 4.56228 8.53678C4.59166 8.55356 4.61972 8.57256 4.64622 8.59361C4.67605 8.6173 4.70318 8.64444 4.75745 8.69871L5.79805 9.7393C5.88293 9.82418 5.92537 9.86662 5.95909 9.91498C5.98903 9.9579 6.0138 10.0042 6.03291 10.0529C6.05444 10.1078 6.06621 10.1666 6.08975 10.2844L6.57835 12.7273C6.70532 13.3622 6.76881 13.6796 6.93618 13.8281C7.08198 13.9575 7.27711 14.0166 7.47019 13.9898C7.69183 13.9591 7.92074 13.7302 8.37855 13.2724L13.1024 8.54856C13.5602 8.09075 13.7891 7.86185 13.8198 7.64021C13.8466 7.44713 13.7875 7.252 13.6581 7.10619C13.5096 6.93882 13.1922 6.87534 12.5573 6.74836L10.1143 6.25977C9.99663 6.23622 9.93777 6.22445 9.88289 6.20293C9.83418 6.18382 9.78788 6.15904 9.74496 6.12911C9.69661 6.09538 9.65416 6.05294 9.56928 5.96806L8.52869 4.92747C8.47442 4.8732 8.44729 4.84606 8.42359 4.81624C8.40255 4.78974 8.38355 4.76168 8.36676 4.73229C8.34787 4.69922 8.33275 4.66395 8.30252 4.5934L7.58984 2.9305Z",
    fill: color
  }), /*#__PURE__*/React__default.createElement("path", {
    d: "M10.7405 10.9105L14.5117 14.6817M8.52869 4.92747L9.56928 5.96806C9.65416 6.05294 9.69661 6.09538 9.74496 6.12911C9.78788 6.15904 9.83418 6.18382 9.88289 6.20293C9.93777 6.22445 9.99663 6.23622 10.1143 6.25977L12.5573 6.74836C13.1922 6.87534 13.5096 6.93882 13.6581 7.10619C13.7875 7.252 13.8466 7.44713 13.8198 7.64021C13.7891 7.86185 13.5602 8.09075 13.1024 8.54856L8.37855 13.2724C7.92074 13.7302 7.69183 13.9591 7.47019 13.9898C7.27711 14.0166 7.08198 13.9575 6.93618 13.8281C6.76881 13.6796 6.70532 13.3622 6.57835 12.7273L6.08975 10.2844C6.06621 10.1666 6.05444 10.1078 6.03291 10.0529C6.0138 10.0042 5.98903 9.9579 5.95909 9.91498C5.92537 9.86662 5.88293 9.82418 5.79805 9.7393L4.75745 8.69871C4.70318 8.64444 4.67605 8.6173 4.64622 8.59361C4.61972 8.57256 4.59166 8.55356 4.56228 8.53678C4.5292 8.51788 4.49393 8.50277 4.42339 8.47253L2.76048 7.75986C2.27535 7.55194 2.03278 7.44799 1.9226 7.28C1.82624 7.13309 1.79177 6.95409 1.82668 6.78191C1.86659 6.58502 2.0532 6.39841 2.42642 6.02519L5.85517 2.59643C6.22839 2.22321 6.415 2.03661 6.6119 1.99669C6.78408 1.96179 6.96308 1.99626 7.10998 2.09261C7.27797 2.2028 7.38193 2.44536 7.58984 2.9305L8.30252 4.5934C8.33275 4.66395 8.34787 4.69922 8.36676 4.73229C8.38355 4.76168 8.40255 4.78974 8.42359 4.81624C8.44729 4.84606 8.47442 4.8732 8.52869 4.92747Z",
    stroke: color,
    strokeWidth: "1.33333",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }));
};

var UnReadIcon = function UnReadIcon(_ref) {
  var style = _ref.style;
  var height = (style === null || style === void 0 ? void 0 : style.height) || 16;
  var width = (style === null || style === void 0 ? void 0 : style.width) || 16;
  var color = (style === null || style === void 0 ? void 0 : style.color) || lightColors.primaryText;
  return /*#__PURE__*/React__default.createElement("svg", {
    width: width,
    height: height,
    viewBox: "0 0 24 24",
    fill: "none",
    xmlns: "http://www.w3.org/2000/svg"
  }, /*#__PURE__*/React__default.createElement("path", {
    fillRule: "evenodd",
    clipRule: "evenodd",
    d: "M14.416 3H4.00013C2.34327 3 1.00013 4.34315 1.00013 6V6.98313C0.999958 6.99375 0.999958 7.00437 1.00013 7.01499V18C1.00013 19.6569 2.34327 21 4.00013 21H20.0001C21.657 21 23.0001 19.6569 23.0001 18V9.21078H21.0001V18C21.0001 18.5523 20.5524 19 20.0001 19H4.00013C3.44784 19 3.00013 18.5523 3.00013 18V8.82027L10.4338 13.544L10.4392 13.5474C10.9071 13.8405 11.448 13.996 12.0001 13.996C12.5522 13.996 13.0932 13.8405 13.5611 13.5474L13.5665 13.544L19.5 9.21078H16.5L12.4992 11.8526L12.497 11.854C12.3479 11.9468 12.1758 11.996 12.0001 11.996C11.8245 11.996 11.6524 11.9468 11.5033 11.854L3.00013 6.45063V6C3.00013 5.44772 3.44784 5 4.00013 5H14.416V3Z",
    fill: color
  }), /*#__PURE__*/React__default.createElement("circle", {
    cx: "20",
    cy: "4",
    r: "4",
    fill: color
  }));
};

var ReadIcon = function ReadIcon(_ref) {
  var style = _ref.style;
  var height = (style === null || style === void 0 ? void 0 : style.height) || 16;
  var width = (style === null || style === void 0 ? void 0 : style.width) || 16;
  var color = (style === null || style === void 0 ? void 0 : style.color) || lightColors.primaryText;
  return /*#__PURE__*/React__default.createElement("svg", {
    width: width,
    height: height,
    viewBox: "0 0 16 17",
    fill: "none",
    xmlns: "http://www.w3.org/2000/svg"
  }, /*#__PURE__*/React__default.createElement("path", {
    d: "M14.1334 6.26471C14.4667 6.51805 14.6667 6.91138 14.6667 7.33138V13.998C14.6667 14.3517 14.5262 14.6908 14.2762 14.9409C14.0261 15.1909 13.687 15.3314 13.3334 15.3314H2.66671C2.31309 15.3314 1.97395 15.1909 1.7239 14.9409C1.47385 14.6908 1.33337 14.3517 1.33337 13.998V7.33138C1.33337 7.12439 1.38157 6.92024 1.47414 6.7351C1.56671 6.54995 1.70111 6.38891 1.86671 6.26471L7.20004 2.26471C7.43084 2.09162 7.71155 1.99805 8.00004 1.99805C8.28853 1.99805 8.56925 2.09162 8.80004 2.26471L14.1334 6.26471Z",
    stroke: color,
    strokeWidth: "1.33333",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React__default.createElement("path", {
    d: "M14.6667 7.33105L8.68668 11.1311C8.48086 11.26 8.24289 11.3284 8.00001 11.3284C7.75713 11.3284 7.51916 11.26 7.31334 11.1311L1.33334 7.33105",
    stroke: color,
    strokeWidth: "1.33333",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }));
};

var ArchiveIcon = function ArchiveIcon(_ref) {
  var style = _ref.style;
  var height = (style === null || style === void 0 ? void 0 : style.height) || 17;
  var width = (style === null || style === void 0 ? void 0 : style.width) || 16;
  var color = (style === null || style === void 0 ? void 0 : style.color) || lightColors.secondaryText;
  return /*#__PURE__*/React__default.createElement("svg", {
    width: width,
    height: height,
    viewBox: "0 0 16 17",
    fill: "none",
    xmlns: "http://www.w3.org/2000/svg"
  }, /*#__PURE__*/React__default.createElement("path", {
    d: "M14 2.33789H1.99998C1.63179 2.33789 1.33331 2.63637 1.33331 3.00456V5.00456C1.33331 5.37275 1.63179 5.67122 1.99998 5.67122H14C14.3682 5.67122 14.6666 5.37275 14.6666 5.00456V3.00456C14.6666 2.63637 14.3682 2.33789 14 2.33789Z",
    stroke: color,
    strokeWidth: "1.33333",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React__default.createElement("path", {
    d: "M2.66669 5.67188V13.0052C2.66669 13.3588 2.80716 13.698 3.05721 13.948C3.30726 14.1981 3.6464 14.3385 4.00002 14.3385H12C12.3536 14.3385 12.6928 14.1981 12.9428 13.948C13.1929 13.698 13.3334 13.3588 13.3334 13.0052V5.67188",
    stroke: color,
    strokeWidth: "1.33333",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React__default.createElement("path", {
    d: "M6.66669 8.33789H9.33335",
    stroke: color,
    strokeWidth: "1.33333",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }));
};

var _templateObject$2, _templateObject2$1, _templateObject3$1, _templateObject4, _templateObject5, _templateObject6, _templateObject7, _templateObject8, _templateObject9, _templateObject10, _templateObject11, _templateObject12, _templateObject13, _templateObject14, _templateObject15, _templateObject16, _templateObject17, _templateObject18, _templateObject19, _templateObject20, _templateObject21, _templateObject22, _templateObject23, _templateObject24, _templateObject25, _templateObject26, _templateObject27;
function ExpiryTime(_ref) {
  var dateInput = _ref.dateInput,
    style = _ref.style;
  var date = dateInput;
  var _useState = React.useState(Date.now()),
    setTime = _useState[1];
  React.useEffect(function () {
    var interval = setInterval(function () {
      return setTime(Date.now());
    }, 10000);
    return function () {
      clearInterval(interval);
    };
  }, []);
  var isExpiring = date - Date.now() <= 3600000;
  var expiredAlready = Date.now() >= date;
  if (typeof dateInput === 'number') {
    return /*#__PURE__*/React__default.createElement("div", null, /*#__PURE__*/React__default.createElement(ExpiresText, {
      style: _extends({}, style, {
        color: isExpiring ? (style === null || style === void 0 ? void 0 : style.expiringColor) || lightColors.error : (style === null || style === void 0 ? void 0 : style.color) || lightColors.secondaryText,
        backgroundColor: isExpiring ? (style === null || style === void 0 ? void 0 : style.expiringBackgroundColor) || 'rgba(217, 45, 32, 0.09)' : (style === null || style === void 0 ? void 0 : style.backgroundColor) || 'rgba(100, 116, 139, 0.09)'
      })
    }, "Expires in", ' ', expiredAlready ? 'a minute' : /*#__PURE__*/React__default.createElement(TimeAgo, {
      date: date,
      live: false,
      formatter: getLongFormattedTime
    })));
  } else {
    return null;
  }
}
function Notification(_ref2) {
  var _notification$bodyTex, _notification$bodyTex2, _message$actions, _message$actions2, _message$avatar2, _message$text, _message$text$replace, _message$text$replace2, _message$subtext, _message$subtext2, _notification$actions, _notification$actions2, _notification$actions3, _notification$actions4, _notification$actions5, _notification$actions6, _notification$actions7, _notification$actions8, _notification$actions9;
  var notificationData = _ref2.notificationData,
    handleActionClick = _ref2.handleActionClick;
  var _useState2 = React.useState(false),
    validAvatar = _useState2[0],
    setValidAvatar = _useState2[1];
  var _useState3 = React.useState(false),
    showMore = _useState3[0],
    setShowMore = _useState3[1];
  var _useState4 = React.useState(false),
    moreOpen = _useState4[0],
    setMoreOpen = _useState4[1];
  var message = notificationData.message,
    seenOn = notificationData.seen_on,
    createdOn = notificationData.created_on;
  var _useInbox = useInbox(),
    notificationComponent = _useInbox.notificationComponent,
    hideAvatar = _useInbox.hideAvatar,
    inbox = _useInbox.inbox,
    themeType = _useInbox.themeType;
  var _useTheme = useTheme(),
    notification = _useTheme.notification;
  var blockquoteColor = (notification === null || notification === void 0 ? void 0 : (_notification$bodyTex = notification.bodyText) === null || _notification$bodyTex === void 0 ? void 0 : _notification$bodyTex.blockquoteColor) || 'rgba(100, 116, 139, 0.3)';
  var linkColor = (notification === null || notification === void 0 ? void 0 : (_notification$bodyTex2 = notification.bodyText) === null || _notification$bodyTex2 === void 0 ? void 0 : _notification$bodyTex2.linkColor) || lightColors.primary;
  var actionOne = message === null || message === void 0 ? void 0 : (_message$actions = message.actions) === null || _message$actions === void 0 ? void 0 : _message$actions[0];
  var actionTwo = message === null || message === void 0 ? void 0 : (_message$actions2 = message.actions) === null || _message$actions2 === void 0 ? void 0 : _message$actions2[1];
  var hasButtons = actionOne || actionTwo;
  React.useEffect(function () {
    var _message$avatar;
    var isValidAvatar = isImgUrl(message === null || message === void 0 ? void 0 : (_message$avatar = message.avatar) === null || _message$avatar === void 0 ? void 0 : _message$avatar.avatar_url);
    isValidAvatar.then(function (res) {
      return setValidAvatar(res);
    });
  }, [notificationData]);
  if (notificationComponent) {
    var NotificationComponent = notificationComponent;
    return /*#__PURE__*/React__default.createElement(NotificationComponent, {
      notificationData: notificationData
    });
  }
  return /*#__PURE__*/React__default.createElement(Container, {
    style: notification === null || notification === void 0 ? void 0 : notification.container,
    read: !!seenOn,
    onMouseEnter: function onMouseEnter() {
      setShowMore(true);
      setMoreOpen(false);
    },
    onMouseLeave: function onMouseLeave() {
      setShowMore(false);
      setMoreOpen(false);
    },
    onClick: function onClick(e) {
      if (moreOpen) {
        e.stopPropagation();
        setMoreOpen(false);
      }
    }
  }, notificationData.is_pinned && /*#__PURE__*/React__default.createElement(PinnedView, {
    hideAvatar: hideAvatar
  }, /*#__PURE__*/React__default.createElement(PinnedNotificationIcon, {
    style: notification === null || notification === void 0 ? void 0 : notification.pinnedIcon
  }), /*#__PURE__*/React__default.createElement(PinnedNotificationText, {
    style: notification === null || notification === void 0 ? void 0 : notification.pinnedText
  }, "Pinned")), /*#__PURE__*/React__default.createElement(NotificationView, null, /*#__PURE__*/React__default.createElement(LeftView, null, /*#__PURE__*/React__default.createElement(LeftAvatarView, null, /*#__PURE__*/React__default.createElement(UnseenDot, {
    style: notification === null || notification === void 0 ? void 0 : notification.unseenDot,
    show: !seenOn
  }), !hideAvatar && /*#__PURE__*/React__default.createElement(AvatarView, {
    onClick: function onClick(e) {
      var avatarData = message === null || message === void 0 ? void 0 : message.avatar;
      handleActionClick(e, {
        type: 'avatar',
        url: avatarData === null || avatarData === void 0 ? void 0 : avatarData.action_url,
        target: avatarData === null || avatarData === void 0 ? void 0 : avatarData.open_in_new_tab
      });
    }
  }, message !== null && message !== void 0 && (_message$avatar2 = message.avatar) !== null && _message$avatar2 !== void 0 && _message$avatar2.avatar_url && validAvatar ? /*#__PURE__*/React__default.createElement(AvatarImage, {
    src: message.avatar.avatar_url,
    alt: "avatar",
    style: notification === null || notification === void 0 ? void 0 : notification.avatar
  }) : /*#__PURE__*/React__default.createElement(AvatarIcon, {
    type: themeType
  }))), /*#__PURE__*/React__default.createElement(ContentView, null, message.header && /*#__PURE__*/React__default.createElement(HeaderText, {
    style: notification === null || notification === void 0 ? void 0 : notification.headerText
  }, message.header), /*#__PURE__*/React__default.createElement(BodyText, {
    style: notification === null || notification === void 0 ? void 0 : notification.bodyText
  }, /*#__PURE__*/React__default.createElement(Markdown, {
    components: {
      a: function a(_ref3) {
        var children = _ref3.children,
          href = _ref3.href;
        return /*#__PURE__*/React__default.createElement("span", {
          onClick: function onClick(e) {
            e.preventDefault();
            handleActionClick(e, {
              type: 'markdown_link',
              url: href
            });
          },
          style: {
            color: linkColor,
            textDecoration: 'none'
          }
        }, children);
      },
      p: function p(_ref4) {
        var children = _ref4.children;
        return /*#__PURE__*/React__default.createElement("p", {
          style: {
            margin: 0,
            overflowWrap: 'anywhere'
          }
        }, children);
      },
      blockquote: function blockquote(_ref5) {
        var children = _ref5.children;
        return /*#__PURE__*/React__default.createElement("blockquote", {
          style: {
            margin: 0,
            paddingLeft: 10,
            borderLeft: "3px " + blockquoteColor + " solid",
            marginBottom: 5,
            marginTop: 5
          }
        }, children);
      },
      ul: function ul(_ref6) {
        var children = _ref6.children;
        return /*#__PURE__*/React__default.createElement("ul", {
          style: {
            whiteSpace: 'normal',
            margin: 0,
            paddingLeft: 10
          }
        }, children);
      },
      ol: function ol(_ref7) {
        var children = _ref7.children;
        return /*#__PURE__*/React__default.createElement("ol", {
          style: {
            whiteSpace: 'normal',
            margin: 0,
            paddingLeft: 10
          }
        }, children);
      },
      img: function img(props) {
        return /*#__PURE__*/React__default.createElement("img", _extends({
          style: {
            maxWidth: '100%',
            objectFit: 'contain'
          }
        }, props));
      }
    }
  }, message === null || message === void 0 ? void 0 : (_message$text = message.text) === null || _message$text === void 0 ? void 0 : (_message$text$replace = _message$text.replaceAll('\\\n', '&nbsp;')) === null || _message$text$replace === void 0 ? void 0 : (_message$text$replace2 = _message$text$replace.replaceAll('\n', '  \n')) === null || _message$text$replace2 === void 0 ? void 0 : _message$text$replace2.replaceAll('&nbsp;', '&nbsp;  \n'))), !!(message !== null && message !== void 0 && (_message$subtext = message.subtext) !== null && _message$subtext !== void 0 && _message$subtext.text) && /*#__PURE__*/React__default.createElement(SubTextView, {
    link: message === null || message === void 0 ? void 0 : (_message$subtext2 = message.subtext) === null || _message$subtext2 === void 0 ? void 0 : _message$subtext2.action_url,
    onClick: function onClick(e) {
      var subTextData = message === null || message === void 0 ? void 0 : message.subtext;
      handleActionClick(e, {
        type: 'subtext',
        url: subTextData === null || subTextData === void 0 ? void 0 : subTextData.action_url,
        target: subTextData === null || subTextData === void 0 ? void 0 : subTextData.open_in_new_tab
      });
    }
  }, /*#__PURE__*/React__default.createElement(SubText, {
    style: notification === null || notification === void 0 ? void 0 : notification.subtext
  }, message.subtext.text)), notificationData.expiry && (notificationData === null || notificationData === void 0 ? void 0 : notificationData.is_expiry_visible) && /*#__PURE__*/React__default.createElement(ExpiryTime, {
    dateInput: notificationData.expiry,
    style: notification === null || notification === void 0 ? void 0 : notification.expiresText
  }), hasButtons && /*#__PURE__*/React__default.createElement(ButtonContainer, null, actionOne && /*#__PURE__*/React__default.createElement(ButtonView, {
    style: notification === null || notification === void 0 ? void 0 : (_notification$actions = notification.actions) === null || _notification$actions === void 0 ? void 0 : (_notification$actions2 = _notification$actions[0]) === null || _notification$actions2 === void 0 ? void 0 : _notification$actions2.container,
    key: actionOne.id,
    onClick: function onClick(e) {
      handleActionClick(e, {
        type: 'action_button',
        url: actionOne.url,
        target: actionOne.open_in_new_tab
      });
    }
  }, /*#__PURE__*/React__default.createElement(ButtonText, {
    style: notification === null || notification === void 0 ? void 0 : (_notification$actions3 = notification.actions) === null || _notification$actions3 === void 0 ? void 0 : (_notification$actions4 = _notification$actions3[0]) === null || _notification$actions4 === void 0 ? void 0 : _notification$actions4.text
  }, actionOne.name)), actionTwo && /*#__PURE__*/React__default.createElement(ButtonOutlineView, {
    key: actionTwo.id,
    style: notification === null || notification === void 0 ? void 0 : (_notification$actions5 = notification.actions) === null || _notification$actions5 === void 0 ? void 0 : (_notification$actions6 = _notification$actions5[1]) === null || _notification$actions6 === void 0 ? void 0 : _notification$actions6.container,
    onClick: function onClick(e) {
      handleActionClick(e, {
        type: 'action_button',
        url: actionTwo.url,
        target: actionTwo.open_in_new_tab
      });
    }
  }, /*#__PURE__*/React__default.createElement(ButtonOutlineText, {
    style: notification === null || notification === void 0 ? void 0 : (_notification$actions7 = notification.actions) === null || _notification$actions7 === void 0 ? void 0 : (_notification$actions8 = _notification$actions7[1]) === null || _notification$actions8 === void 0 ? void 0 : _notification$actions8.text
  }, actionTwo.name))))), /*#__PURE__*/React__default.createElement(RightView, null, /*#__PURE__*/React__default.createElement(CreatedText, {
    style: notification === null || notification === void 0 ? void 0 : notification.createdOnText
  }, /*#__PURE__*/React__default.createElement(TimeAgo, {
    date: createdOn,
    live: false,
    formatter: getShortFormattedTime
  })), /*#__PURE__*/React__default.createElement(CMenuView, {
    showMore: showMore
  }, /*#__PURE__*/React__default.createElement(CMenuButton, {
    hoverBGColor: notification === null || notification === void 0 ? void 0 : (_notification$actions9 = notification.actionsMenuIcon) === null || _notification$actions9 === void 0 ? void 0 : _notification$actions9.hoverBackgroundColor,
    onClick: function onClick(e) {
      e.stopPropagation();
      setMoreOpen(function (prev) {
        return !prev;
      });
    }
  }, /*#__PURE__*/React__default.createElement(MoreIcon, {
    style: notification === null || notification === void 0 ? void 0 : notification.actionsMenuIcon
  })), /*#__PURE__*/React__default.createElement(CMenuPopup, {
    moreOpen: moreOpen,
    style: notification === null || notification === void 0 ? void 0 : notification.actionsMenu
  }, notificationData.seen_on ? /*#__PURE__*/React__default.createElement(CMenuItem, {
    style: notification === null || notification === void 0 ? void 0 : notification.actionsMenuItem,
    onClick: function onClick(e) {
      e.stopPropagation();
      inbox.feed.markUnRead(notificationData.n_id);
      setMoreOpen(false);
    }
  }, /*#__PURE__*/React__default.createElement(UnReadIcon, {
    style: notification === null || notification === void 0 ? void 0 : notification.actionsMenuItemIcon
  }), /*#__PURE__*/React__default.createElement(CMenuText, {
    style: notification === null || notification === void 0 ? void 0 : notification.actionsMenuItemText
  }, "Mark as unread")) : /*#__PURE__*/React__default.createElement(CMenuItem, {
    style: notification === null || notification === void 0 ? void 0 : notification.actionsMenuItem,
    onClick: function onClick(e) {
      e.stopPropagation();
      inbox.feed.markRead(notificationData.n_id);
      setMoreOpen(false);
    }
  }, /*#__PURE__*/React__default.createElement(ReadIcon, {
    style: notification === null || notification === void 0 ? void 0 : notification.actionsMenuItemIcon
  }), /*#__PURE__*/React__default.createElement(CMenuText, {
    style: notification === null || notification === void 0 ? void 0 : notification.actionsMenuItemText
  }, "Mark as read")), !(notificationData !== null && notificationData !== void 0 && notificationData.archived) && /*#__PURE__*/React__default.createElement(CMenuItem, {
    style: notification === null || notification === void 0 ? void 0 : notification.actionsMenuItem,
    onClick: function onClick(e) {
      e.stopPropagation();
      inbox.feed.markArchived(notificationData.n_id);
    }
  }, /*#__PURE__*/React__default.createElement(ArchiveIcon, {
    style: notification === null || notification === void 0 ? void 0 : notification.actionsMenuItemIcon
  }), /*#__PURE__*/React__default.createElement(CMenuText, {
    style: notification === null || notification === void 0 ? void 0 : notification.actionsMenuItemText
  }, "Archive")))))));
}
var Container = styled.div(_templateObject$2 || (_templateObject$2 = _taggedTemplateLiteralLoose(["\n  padding: 16px;\n  cursor: pointer;\n  background-color: ", ";\n  border-bottom: 0.5px solid ", ";\n  &:hover {\n    background-color: ", "\n"])), function (props) {
  var _props$style, _props$style2;
  return props.read ? (props === null || props === void 0 ? void 0 : (_props$style = props.style) === null || _props$style === void 0 ? void 0 : _props$style.readBackgroundColor) || lightColors.main : (props === null || props === void 0 ? void 0 : (_props$style2 = props.style) === null || _props$style2 === void 0 ? void 0 : _props$style2.unreadBackgroundColor) || '#edf3ff';
}, lightColors.border, function (props) {
  var _props$style3;
  return (props === null || props === void 0 ? void 0 : (_props$style3 = props.style) === null || _props$style3 === void 0 ? void 0 : _props$style3.hoverBackgroundColor) || '#DBE7FF';
});
var PinnedView = styled.div(_templateObject2$1 || (_templateObject2$1 = _taggedTemplateLiteralLoose(["\n  display: flex;\n  align-items: center;\n  margin-left: ", ";\n  gap: 4px;\n"])), function (props) {
  return props.hideAvatar ? '0px' : '52px';
});
var PinnedNotificationText = styled(HelperText)(_templateObject3$1 || (_templateObject3$1 = _taggedTemplateLiteralLoose([""])));
var SubText = styled(HelperText)(_templateObject4 || (_templateObject4 = _taggedTemplateLiteralLoose(["\n  color: #64748b;\n"])));
var SubTextView = styled.div(_templateObject5 || (_templateObject5 = _taggedTemplateLiteralLoose(["\n  text-decoration: none;\n  overflow-wrap: anywhere;\n  display: inline-block;\n  &:hover {\n    text-decoration: ", ";\n    text-decoration-color: ", ";\n  }\n"])), function (props) {
  return props.link ? 'underline' : 'none';
}, lightColors.secondaryText);
var ExpiresText = styled(HelperText)(_templateObject6 || (_templateObject6 = _taggedTemplateLiteralLoose(["\n  margin-top: 8px;\n  display: inline-block;\n  padding: 1px 6px;\n  border-radius: 4px;\n"])));
var NotificationView = styled.div(_templateObject7 || (_templateObject7 = _taggedTemplateLiteralLoose(["\n  display: flex;\n  flex-direction: row;\n  align-items: flex-start;\n  justify-content: space-between;\n"])));
var HeaderText = styled(CText)(_templateObject8 || (_templateObject8 = _taggedTemplateLiteralLoose(["\n  margin: 8px 0px 0px 0px;\n  overflow-wrap: anywhere;\n  line-height: 18px;\n  font-weight: 700;\n"])));
var BodyText = styled.div(_templateObject9 || (_templateObject9 = _taggedTemplateLiteralLoose(["\n  font-size: 14px;\n  line-height: 20px;\n  margin: 8px 0px 6px;\n  font-weight: 400;\n  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Helvetica,\n    Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol';\n  color: ", ";\n"])), lightColors.secondaryText);
var UnseenDot = styled.div(_templateObject10 || (_templateObject10 = _taggedTemplateLiteralLoose(["\n  background-color: ", ";\n  border-radius: 50%;\n  width: 7px;\n  height: 7px;\n  visibility: ", ";\n"])), lightColors.primary, function (props) {
  return props.show ? 'visible' : 'hidden';
});
var CreatedText = styled(HelperText)(_templateObject11 || (_templateObject11 = _taggedTemplateLiteralLoose([""])));
var ButtonContainer = styled.div(_templateObject12 || (_templateObject12 = _taggedTemplateLiteralLoose(["\n  display: flex;\n  flex-direction: row;\n  gap: 10px;\n  margin-bottom: 6px;\n  margin-top: 12px;\n  overflow-wrap: anywhere;\n"])));
var ButtonView = styled.div(_templateObject13 || (_templateObject13 = _taggedTemplateLiteralLoose(["\n  max-width: 50%;\n  background: ", ";\n  border-radius: 5px;\n  text-decoration: none;\n  padding: 4px 16px;\n  &:hover {\n    background-color: ", " !important;\n  }\n"])), lightColors.primary, function (props) {
  var _props$style4;
  return ((_props$style4 = props.style) === null || _props$style4 === void 0 ? void 0 : _props$style4.hoverBackgroundColor) || '#265cbf';
});
var ButtonText = styled(CText)(_templateObject14 || (_templateObject14 = _taggedTemplateLiteralLoose(["\n  color: ", ";\n  text-align: center;\n  word-break: break-all;\n  font-weight: 600;\n  font-size: 13px;\n"])), lightColors.main);
var ButtonOutlineView = styled(ButtonView)(_templateObject15 || (_templateObject15 = _taggedTemplateLiteralLoose(["\n  background: ", ";\n  border-color: ", ";\n  border-style: solid;\n  border-width: 1.2px;\n  &:hover {\n    background-color: ", " !important;\n  }\n"])), lightColors.main, lightColors.border, function (props) {
  var _props$style5;
  return ((_props$style5 = props.style) === null || _props$style5 === void 0 ? void 0 : _props$style5.hoverBackgroundColor) || '#f7f7f9';
});
var ButtonOutlineText = styled(ButtonText)(_templateObject16 || (_templateObject16 = _taggedTemplateLiteralLoose(["\n  color: ", ";\n"])), lightColors.secondaryText);
var LeftView = styled.div(_templateObject17 || (_templateObject17 = _taggedTemplateLiteralLoose(["\n  display: flex;\n  overflow-wrap: anywhere;\n  flex-grow: 1;\n"])));
var LeftAvatarView = styled.div(_templateObject18 || (_templateObject18 = _taggedTemplateLiteralLoose(["\n  display: flex;\n  margin-top: 8px;\n  margin-right: 12px;\n"])));
var AvatarView = styled.div(_templateObject19 || (_templateObject19 = _taggedTemplateLiteralLoose([""])));
var ContentView = styled.div(_templateObject20 || (_templateObject20 = _taggedTemplateLiteralLoose(["\n  flex: 1;\n"])));
var RightView = styled.div(_templateObject21 || (_templateObject21 = _taggedTemplateLiteralLoose(["\n  display: flex;\n  flex-direction: column;\n  align-items: center;\n  margin-top: 5px;\n  max-width: 40px;\n  gap: 5px;\n"])));
var AvatarImage = styled.img(_templateObject22 || (_templateObject22 = _taggedTemplateLiteralLoose(["\n  height: 36px;\n  width: 36px;\n  border-radius: 100px;\n"])));
var CMenuPopup = styled.div(_templateObject23 || (_templateObject23 = _taggedTemplateLiteralLoose(["\n  position: absolute;\n  right: 0px;\n  display: ", ";\n  min-width: 150px;\n  padding: 2px;\n  background-color: ", ";\n  border: 1px solid;\n  border-color: ", ";\n  border-radius: 4px;\n  box-shadow: 1px 1px 20px 1px rgba(0, 0, 0, 0.1);\n  z-index: 100;\n"])), function (props) {
  return props.moreOpen ? 'block' : 'none';
}, lightColors.main, lightColors.border);
var CMenuItem = styled.div(_templateObject24 || (_templateObject24 = _taggedTemplateLiteralLoose(["\n  padding: 7px;\n  border-radius: 4px;\n  display: flex;\n  align-items: center;\n  gap: 8px;\n  &:hover {\n    background-color: ", ";\n  }\n"])), function (props) {
  var _props$style6;
  return (props === null || props === void 0 ? void 0 : (_props$style6 = props.style) === null || _props$style6 === void 0 ? void 0 : _props$style6.hoverBackgroundColor) || 'rgba(100, 116, 139, 0.09)';
});
var CMenuText = styled(CText)(_templateObject25 || (_templateObject25 = _taggedTemplateLiteralLoose([""])));
var CMenuView = styled.div(_templateObject26 || (_templateObject26 = _taggedTemplateLiteralLoose(["\n  position: relative;\n  visibility: ", ";\n"])), function (props) {
  return props !== null && props !== void 0 && props.showMore ? 'visible' : 'hidden';
});
var CMenuButton = styled.div(_templateObject27 || (_templateObject27 = _taggedTemplateLiteralLoose(["\n  height: 20px;\n  width: 20px;\n  &:hover {\n    border-radius: 50%;\n    background-color: ", ";\n  }\n"])), function (props) {
  return (props === null || props === void 0 ? void 0 : props.hoverBGColor) || 'rgba(100, 116, 139, 0.09)';
});

function getURLTarget(target) {
  return target ? '_blank' : '_self';
}
function ClickableNotification(_ref) {
  var notificationData = _ref.notificationData;
  var _useInbox = useInbox(),
    notificationClickHandler = _useInbox.notificationClickHandler,
    inbox = _useInbox.inbox;
  var cardClickNavigation = function cardClickNavigation() {
    if (typeof notificationClickHandler === 'function') {
      notificationClickHandler(notificationData);
    } else {
      var message = notificationData === null || notificationData === void 0 ? void 0 : notificationData.message;
      if (message !== null && message !== void 0 && message.url) {
        var actionUrlTarget = getURLTarget(message === null || message === void 0 ? void 0 : message.open_in_new_tab);
        window.open(formatActionLink(message.url), actionUrlTarget);
      }
    }
  };
  var markNotificationClicked = function markNotificationClicked() {
    inbox.feed.markClicked(notificationData.n_id);
  };
  var handleCardClick = function handleCardClick(e) {
    var _notificationData$mes;
    e.stopPropagation();
    var clicked = notificationData.interacted_on;
    var actionUrlTarget = getURLTarget(notificationData === null || notificationData === void 0 ? void 0 : (_notificationData$mes = notificationData.message) === null || _notificationData$mes === void 0 ? void 0 : _notificationData$mes.open_in_new_tab);
    markNotificationClicked();
    if (!clicked && actionUrlTarget === '_self') {
      setTimeout(function () {
        cardClickNavigation();
      }, 1000);
    } else {
      cardClickNavigation();
    }
  };
  var handleActionClick = function handleActionClick(e, clickData) {
    e.stopPropagation();
    var clicked = notificationData.interacted_on;
    var actionUrlTarget = getURLTarget(clickData === null || clickData === void 0 ? void 0 : clickData.target);
    markNotificationClicked();
    if (!clicked && actionUrlTarget === '_self') {
      setTimeout(function () {
        if (clickData !== null && clickData !== void 0 && clickData.url) {
          window.open(formatActionLink(clickData.url), actionUrlTarget);
        } else {
          cardClickNavigation();
        }
      }, 1000);
    } else {
      if (clickData !== null && clickData !== void 0 && clickData.url) {
        window.open(formatActionLink(clickData.url), actionUrlTarget);
      } else {
        cardClickNavigation();
      }
    }
  };
  return /*#__PURE__*/React__default.createElement("div", {
    onClick: handleCardClick
  }, /*#__PURE__*/React__default.createElement(Notification, {
    notificationData: notificationData,
    handleActionClick: handleActionClick
  }));
}

var EmptyNotificationIcon = function EmptyNotificationIcon(props) {
  return /*#__PURE__*/React.createElement("svg", _extends({
    width: 185,
    height: 114,
    fill: "none",
    xmlns: "http://www.w3.org/2000/svg"
  }, props), /*#__PURE__*/React.createElement("ellipse", {
    cx: 92.32,
    cy: 56.616,
    rx: 60.32,
    ry: 56.616,
    fill: "#9F9F9F",
    fillOpacity: 0.11
  }), /*#__PURE__*/React.createElement("rect", {
    x: 69.174,
    y: 17.469,
    width: 115.826,
    height: 40.217,
    rx: 12.87,
    fill: "#9F9F9F"
  }), /*#__PURE__*/React.createElement("rect", {
    x: 123.065,
    y: 29.533,
    width: 33.783,
    height: 5.63,
    rx: 2.815,
    fill: "#D9D9D9"
  }), /*#__PURE__*/React.createElement("rect", {
    x: 95.717,
    y: 39.185,
    width: 59.522,
    height: 5.63,
    rx: 2.815,
    fill: "#D9D9D9"
  }), /*#__PURE__*/React.createElement("ellipse", {
    cx: 170.924,
    cy: 37.577,
    rx: 8.446,
    ry: 8.043,
    fill: "#D9D9D9"
  }), /*#__PURE__*/React.createElement("rect", {
    y: 51.25,
    width: 120.652,
    height: 40.217,
    rx: 12.87,
    fill: "#D9E9FF"
  }), /*#__PURE__*/React.createElement("rect", {
    x: 30.565,
    y: 63.315,
    width: 33.783,
    height: 5.63,
    rx: 2.815,
    fill: "#93AFD6"
  }), /*#__PURE__*/React.createElement("rect", {
    x: 30.565,
    y: 72.967,
    width: 59.522,
    height: 5.63,
    rx: 2.815,
    fill: "#93AFD6"
  }), /*#__PURE__*/React.createElement("ellipse", {
    cx: 17.294,
    cy: 71.358,
    rx: 8.446,
    ry: 8.043,
    fill: "#93AFD6"
  }));
};

var _templateObject$3, _templateObject2$2, _templateObject3$2, _templateObject4$1, _templateObject5$1, _templateObject6$1, _templateObject7$1, _templateObject8$1;
function Loader(_ref) {
  var style = _ref.style,
    size = _ref.size;
  var color = style === null || style === void 0 ? void 0 : style.color;
  return /*#__PURE__*/React__default.createElement(SpinnerContainer, null, /*#__PURE__*/React__default.createElement(Spinner, {
    style: style,
    color: color,
    size: size
  }));
}
function NotificationsList() {
  var _storeData$activeStor;
  var _useInbox = useInbox(),
    _useInbox$notificatio = _useInbox.notificationsData,
    storeData = _useInbox$notificatio.storeData,
    hasNext = _useInbox$notificatio.hasNext,
    initialLoading = _useInbox$notificatio.initialLoading,
    activeStoreId = _useInbox$notificatio.activeStoreId,
    noNotificationsComponent = _useInbox.noNotificationsComponent,
    loaderComponent = _useInbox.loaderComponent,
    pagination = _useInbox.pagination,
    changingActiveStore = _useInbox.changingActiveStore,
    inbox = _useInbox.inbox;
  var _useTheme = useTheme(),
    notificationsContainer = _useTheme.notificationsContainer;
  var notifications = (storeData === null || storeData === void 0 ? void 0 : (_storeData$activeStor = storeData[activeStoreId]) === null || _storeData$activeStor === void 0 ? void 0 : _storeData$activeStor.notifications) || [];
  if (initialLoading) {
    return /*#__PURE__*/React__default.createElement(InitialLoader, null, /*#__PURE__*/React__default.createElement(Loader, {
      style: notificationsContainer === null || notificationsContainer === void 0 ? void 0 : notificationsContainer.loader,
      size: "large"
    }));
  }
  if (notifications.length <= 0) {
    if (noNotificationsComponent) {
      var NoNotificationsComponent = noNotificationsComponent;
      return /*#__PURE__*/React__default.createElement(NoNotificationsComponent, null);
    }
    return /*#__PURE__*/React__default.createElement(EmptyNotificationsContainer, null, /*#__PURE__*/React__default.createElement(EmptyNotificationIcon, null), /*#__PURE__*/React__default.createElement(EmptyText, {
      style: notificationsContainer === null || notificationsContainer === void 0 ? void 0 : notificationsContainer.noNotificationsText
    }, "No notifications yet"), /*#__PURE__*/React__default.createElement(EmptySubText, {
      style: notificationsContainer === null || notificationsContainer === void 0 ? void 0 : notificationsContainer.noNotificationsSubtext
    }, "We'll let you know when we've got something new for you."));
  }
  if (changingActiveStore) {
    return null;
  }
  return /*#__PURE__*/React__default.createElement(ScrollDiv, null, /*#__PURE__*/React__default.createElement(InfiniteScroll, {
    scrollableTarget: "ss-notification-container",
    dataLength: notifications.length,
    next: function next() {
      inbox.feed.fetchNotifications();
    },
    hasMore: pagination && hasNext,
    loader: (loaderComponent === null || loaderComponent === void 0 ? void 0 : loaderComponent()) || /*#__PURE__*/React__default.createElement(Loader, {
      style: notificationsContainer === null || notificationsContainer === void 0 ? void 0 : notificationsContainer.loader
    })
  }, notifications.map(function (notification) {
    return /*#__PURE__*/React__default.createElement(ClickableNotification, {
      notificationData: notification,
      key: notification.n_id
    });
  })));
}
var ScrollDiv = styled.div(_templateObject$3 || (_templateObject$3 = _taggedTemplateLiteralLoose([""])));
var EmptyNotificationsContainer = styled.div(_templateObject2$2 || (_templateObject2$2 = _taggedTemplateLiteralLoose(["\n  display: flex;\n  justify-content: center;\n  align-items: center;\n  flex-direction: column;\n  margin: 32px;\n  margin-top: 40px;\n"])));
var EmptyText = styled(HeadingText)(_templateObject3$2 || (_templateObject3$2 = _taggedTemplateLiteralLoose(["\n  font-weight: 600;\n  text-align: center;\n  background-color: transparent;\n  margin: 20px 0px;\n  margin-bottom: 16px;\n  color: ", ";\n"])), lightColors.primaryText);
var EmptySubText = styled(CText)(_templateObject4$1 || (_templateObject4$1 = _taggedTemplateLiteralLoose(["\n  color: ", ";\n  text-align: center;\n"])), lightColors.secondaryText);
var spin = react.keyframes(_templateObject5$1 || (_templateObject5$1 = _taggedTemplateLiteralLoose(["\n0% {\n  transform: rotate(0deg);\n}\n100% {\n  transform: rotate(360deg);\n}\n"])));
var Spinner = styled.div(_templateObject6$1 || (_templateObject6$1 = _taggedTemplateLiteralLoose(["\n  border: ", ";\n  border-top: ", ";\n  border-top-color: ", ";\n  border-radius: 50%;\n  width: ", ";\n  height: ", ";\n  animation: ", " 1s linear infinite;\n  margin: 5px;\n"])), function (props) {
  return props.size === 'large' ? "5px solid " + lightColors.border : "3px solid " + lightColors.border;
}, function (props) {
  return props.size === 'large' ? '5px solid' : '3px solid';
}, function (props) {
  return props.color ? props.color : lightColors.primary;
}, function (props) {
  return props.size === 'large' ? '20px' : '10px';
}, function (props) {
  return props.size === 'large' ? '20px' : '10px';
}, spin);
var SpinnerContainer = styled.div(_templateObject7$1 || (_templateObject7$1 = _taggedTemplateLiteralLoose(["\n  display: flex;\n  justify-content: center;\n  align-items: center;\n"])));
var InitialLoader = styled.div(_templateObject8$1 || (_templateObject8$1 = _taggedTemplateLiteralLoose(["\n  margin-top: 32px;\n"])));

var _templateObject$4, _templateObject2$3, _templateObject3$3, _templateObject4$2, _templateObject5$2, _templateObject6$2, _templateObject7$2, _templateObject8$2, _templateObject9$1;
function hasNotifications(storeData) {
  var hasData = false;
  for (var storeId in storeData) {
    var _store$notifications;
    var store = storeData[storeId];
    if ((store === null || store === void 0 ? void 0 : (_store$notifications = store.notifications) === null || _store$notifications === void 0 ? void 0 : _store$notifications.length) > 0) {
      hasData = true;
    }
  }
  return hasData;
}
function Header() {
  var _useTheme = useTheme(),
    header = _useTheme.header,
    tabs = _useTheme.tabs;
  var _useInbox = useInbox(),
    inbox = _useInbox.inbox,
    notificationsData = _useInbox.notificationsData,
    activeStore = _useInbox.activeStore,
    setActiveStore = _useInbox.setActiveStore,
    setChangingActiveStore = _useInbox.setChangingActiveStore,
    showUnreadCountOnTabs = _useInbox.showUnreadCountOnTabs,
    tabBadgeComponent = _useInbox.tabBadgeComponent;
  var isEmpty = !hasNotifications(notificationsData === null || notificationsData === void 0 ? void 0 : notificationsData.storeData);
  var stores = inbox.feed.stores;
  var hasStores = (stores === null || stores === void 0 ? void 0 : stores.length) > 0;
  var TabBadgeComponent = tabBadgeComponent;
  return /*#__PURE__*/React__default.createElement(Container$1, {
    style: header === null || header === void 0 ? void 0 : header.container
  }, /*#__PURE__*/React__default.createElement(TopContainer, {
    hasStores: hasStores
  }, /*#__PURE__*/React__default.createElement(HeaderText$1, {
    style: header === null || header === void 0 ? void 0 : header.headerText
  }, "Notifications"), !isEmpty && /*#__PURE__*/React__default.createElement(AllReadButton, {
    style: header === null || header === void 0 ? void 0 : header.markAllReadText,
    onClick: function onClick(e) {
      e.stopPropagation();
      inbox.feed.markAllRead();
    }
  }, "Mark all as read")), stores && stores.length > 0 && /*#__PURE__*/React__default.createElement(TabsContainer, null, stores.map(function (store, index) {
    var _notificationsData$st, _notificationsData$st2;
    var isActiveTab = activeStore === store.storeId;
    var tabUnseenCount = (notificationsData === null || notificationsData === void 0 ? void 0 : (_notificationsData$st = notificationsData.storeData) === null || _notificationsData$st === void 0 ? void 0 : (_notificationsData$st2 = _notificationsData$st[store.storeId]) === null || _notificationsData$st2 === void 0 ? void 0 : _notificationsData$st2.unseenCount) || 0;
    var showBadge = showUnreadCountOnTabs && tabUnseenCount > 0;
    var selectedTabBottomColor = isActiveTab ? tabs === null || tabs === void 0 ? void 0 : tabs.bottomColor : 'none';
    var textColor = isActiveTab ? tabs === null || tabs === void 0 ? void 0 : tabs.color : (tabs === null || tabs === void 0 ? void 0 : tabs.unselectedColor) || (tabs === null || tabs === void 0 ? void 0 : tabs.color);
    return /*#__PURE__*/React__default.createElement(TabContainer, {
      style: {
        borderBottomColor: selectedTabBottomColor
      },
      key: index,
      selected: isActiveTab,
      onClick: function onClick() {
        setChangingActiveStore(true);
        setActiveStore(store.storeId);
        inbox.changeActiveStore(store.storeId);
        setTimeout(function () {
          setChangingActiveStore(false);
        }, 0);
      }
    }, /*#__PURE__*/React__default.createElement(TabText, {
      selected: isActiveTab,
      style: _extends({}, tabs, {
        color: textColor
      })
    }, store.label), showBadge && (TabBadgeComponent ? /*#__PURE__*/React__default.createElement(TabBadgeComponent, {
      count: tabUnseenCount
    }) : /*#__PURE__*/React__default.createElement(TabBadge, {
      style: {
        backgroundColor: tabs === null || tabs === void 0 ? void 0 : tabs.badgeColor,
        color: tabs === null || tabs === void 0 ? void 0 : tabs.badgeText
      }
    }, /*#__PURE__*/React__default.createElement(TabBadgeText, {
      count: tabUnseenCount
    }, tabUnseenCount))));
  })));
}
var Container$1 = styled.div(_templateObject$4 || (_templateObject$4 = _taggedTemplateLiteralLoose(["\n  position: sticky;\n  top: 0;\n  padding: 16px;\n  padding-bottom: 0px;\n  z-index: 1000;\n  background-color: ", ";\n  box-shadow: 0 0 3px 0 rgba(0, 0, 0, 0.1);\n"])), lightColors.main);
var TopContainer = styled.div(_templateObject2$3 || (_templateObject2$3 = _taggedTemplateLiteralLoose(["\n  display: flex;\n  justify-content: space-between;\n  align-items: center;\n  margin-bottom: ", ";\n  padding-bottom: ", ";\n"])), function (props) {
  return props.hasStores ? '16px' : '0px';
}, function (props) {
  return props.hasStores ? '0px' : '16px';
});
var TabsContainer = styled.div(_templateObject3$3 || (_templateObject3$3 = _taggedTemplateLiteralLoose(["\n  display: flex;\n  overflow-x: scroll;\n  -ms-overflow-style: none;\n  scrollbar-width: none;\n  &::-webkit-scrollbar {\n    display: none;\n  }\n"])));
var TabContainer = styled.div(_templateObject4$2 || (_templateObject4$2 = _taggedTemplateLiteralLoose(["\n  padding: 5px 15px 3px 15px;\n  border-bottom: ", ";\n  cursor: pointer;\n  display: flex;\n  align-items: center;\n  gap: 5px;\n"])), function (props) {
  return props.selected ? "2px solid " + lightColors.primary : 'none';
});
var TabText = styled(CText)(_templateObject5$2 || (_templateObject5$2 = _taggedTemplateLiteralLoose(["\n  font-size: 14px;\n  font-weight: 600;\n  color: ", ";\n"])), function (props) {
  return props.selected ? lightColors.primaryText : lightColors.secondaryText + 'D9';
});
var HeaderText$1 = styled(HeadingText)(_templateObject6$2 || (_templateObject6$2 = _taggedTemplateLiteralLoose(["\n  font-weight: 600;\n  font-size: 16px;\n"])));
var AllReadButton = styled(HeadingText)(_templateObject7$2 || (_templateObject7$2 = _taggedTemplateLiteralLoose(["\n  color: ", ";\n  font-size: 12px;\n  cursor: pointer;\n"])), lightColors.primary);
var TabBadge = styled.div(_templateObject8$2 || (_templateObject8$2 = _taggedTemplateLiteralLoose(["\n  height: 18px;\n  width: 20px;\n  border-radius: 50%;\n  background-color: rgba(100, 116, 139, 0.09);\n  display: flex;\n  align-items: center;\n  justify-content: center;\n  color: ", ";\n"])), lightColors.primaryText);
var TabBadgeText = styled.p(_templateObject9$1 || (_templateObject9$1 = _taggedTemplateLiteralLoose(["\n  margin: 0px;\n  padding: 0px;\n  font-size: ", ";\n  font-weight: 600;\n  line-height: 1;\n  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Helvetica,\n    Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol';\n"])), function (props) {
  return (props === null || props === void 0 ? void 0 : props.count) > 99 ? '8px' : '10px';
});

var _templateObject$5;
function NotificationsContainer() {
  var _useTheme = useTheme(),
    notificationsContainer = _useTheme.notificationsContainer;
  return /*#__PURE__*/React__default.createElement(Container$2, {
    style: notificationsContainer === null || notificationsContainer === void 0 ? void 0 : notificationsContainer.container,
    id: "ss-notification-container"
  }, /*#__PURE__*/React__default.createElement(Header, null), /*#__PURE__*/React__default.createElement(NotificationsList, null));
}
var Container$2 = styled.div(_templateObject$5 || (_templateObject$5 = _taggedTemplateLiteralLoose(["\n  height: 500px;\n  width: 450px;\n  margin: 0px 15px;\n  border-radius: 5px;\n  background-color: ", ";\n  border: 1px solid ", ";\n  display: inline-block;\n  overflow: scroll;\n  box-shadow: 0 0px 7px 0 rgba(0, 0, 0, 0.1);\n  @media (max-width: 425px) {\n    width: 99.5vw;\n    margin: 0px;\n    border-radius: 0px;\n  }\n"])), lightColors.main, lightColors.border);

var useClickOutside = function useClickOutside(ref, handler) {
  React.useEffect(function () {
    var startedInside = false;
    var startedWhenMounted = false;
    var listener = function listener(event) {
      if (startedInside || !startedWhenMounted) return;
      if (!ref.current || ref.current.contains(event.target)) return;
      handler(event);
    };
    var validateEventStart = function validateEventStart(event) {
      startedWhenMounted = ref.current;
      startedInside = ref.current && ref.current.contains(event.target);
    };
    document.addEventListener('mousedown', validateEventStart);
    document.addEventListener('touchstart', validateEventStart);
    document.addEventListener('click', listener);
    return function () {
      document.removeEventListener('mousedown', validateEventStart);
      document.removeEventListener('touchstart', validateEventStart);
      document.removeEventListener('click', listener);
    };
  }, [ref, handler]);
};

var _templateObject$6, _templateObject2$4;
function Inbox(_ref) {
  var openInbox = _ref.openInbox,
    toggleInbox = _ref.toggleInbox,
    popperPosition = _ref.popperPosition;
  var _useState = React.useState(null),
    referenceElement = _useState[0],
    setReferenceElement = _useState[1];
  var _useState2 = React.useState(null),
    popperElement = _useState2[0],
    setPopperElement = _useState2[1];
  var _useInbox = useInbox(),
    setNotificationsData = _useInbox.setNotificationsData;
  useClickOutside({
    current: popperElement
  }, function () {
    toggleInbox(function (prev) {
      return !prev;
    });
  });
  var _usePopper = reactPopper.usePopper(referenceElement, popperElement, {
      placement: popperPosition,
      modifiers: [{
        name: 'offset',
        options: {
          offset: [0, 5]
        }
      }]
    }),
    styles = _usePopper.styles,
    attributes = _usePopper.attributes;
  var handleBellClick = function handleBellClick() {
    setNotificationsData(function (prev) {
      return _extends({}, prev, {
        count: 0
      });
    });
    toggleInbox(function (prev) {
      return !prev;
    });
  };
  return /*#__PURE__*/React__default.createElement(Container$3, null, /*#__PURE__*/React__default.createElement(BellContainer, {
    onClick: handleBellClick,
    ref: setReferenceElement
  }, /*#__PURE__*/React__default.createElement(Badge, null), /*#__PURE__*/React__default.createElement(Bell, null)), openInbox && /*#__PURE__*/React__default.createElement("div", _extends({
    ref: setPopperElement,
    style: _extends({}, styles.popper, {
      zIndex: 999
    })
  }, attributes.popper), /*#__PURE__*/React__default.createElement(NotificationsContainer, null)));
}
var Container$3 = styled.div(_templateObject$6 || (_templateObject$6 = _taggedTemplateLiteralLoose(["\n  position: relative;\n  display: inline-block;\n  background-color: transparent;\n  line-height: 1;\n"])));
var BellContainer = styled.div(_templateObject2$4 || (_templateObject2$4 = _taggedTemplateLiteralLoose(["\n  position: relative;\n  margin-top: 12px;\n  margin-right: 12px;\n  cursor: pointer;\n"])));

var _templateObject$7, _templateObject2$5, _templateObject3$4, _templateObject4$3, _templateObject5$3;
function ToastNotification(_ref) {
  var _toast$bodyText, _toast$bodyText2, _message$avatar2, _message$text, _message$text$replace, _message$text$replace2;
  var notificationData = _ref.notificationData,
    closeToast = _ref.closeToast;
  var _useTheme = useTheme(),
    toast = _useTheme.toast;
  var _useInbox = useInbox(),
    toggleInbox = _useInbox.toggleInbox,
    themeType = _useInbox.themeType;
  var message = notificationData.message;
  var _useState = React.useState(false),
    validAvatar = _useState[0],
    setValidAvatar = _useState[1];
  React.useEffect(function () {
    var _message$avatar;
    var isValidAvatar = isImgUrl(message === null || message === void 0 ? void 0 : (_message$avatar = message.avatar) === null || _message$avatar === void 0 ? void 0 : _message$avatar.avatar_url);
    isValidAvatar.then(function (res) {
      return setValidAvatar(res);
    });
  }, [notificationData]);
  var blockquoteColor = (toast === null || toast === void 0 ? void 0 : (_toast$bodyText = toast.bodyText) === null || _toast$bodyText === void 0 ? void 0 : _toast$bodyText.blockquoteColor) || lightColors.border;
  var linkColor = (toast === null || toast === void 0 ? void 0 : (_toast$bodyText2 = toast.bodyText) === null || _toast$bodyText2 === void 0 ? void 0 : _toast$bodyText2.linkColor) || lightColors.primary;
  return /*#__PURE__*/React__default.createElement(Container$4, {
    style: toast === null || toast === void 0 ? void 0 : toast.container,
    onClick: function onClick(e) {
      toggleInbox(true);
      closeToast();
    }
  }, /*#__PURE__*/React__default.createElement(AvatarView$1, null, message !== null && message !== void 0 && (_message$avatar2 = message.avatar) !== null && _message$avatar2 !== void 0 && _message$avatar2.avatar_url && validAvatar ? /*#__PURE__*/React__default.createElement(AvatarImage$1, {
    src: message.avatar.avatar_url,
    alt: "avatar"
  }) : /*#__PURE__*/React__default.createElement(AvatarIcon, {
    type: themeType
  })), /*#__PURE__*/React__default.createElement("div", null, /*#__PURE__*/React__default.createElement(HeaderText$2, {
    style: toast === null || toast === void 0 ? void 0 : toast.headerText
  }, message.header), /*#__PURE__*/React__default.createElement(BodyText$1, {
    style: toast === null || toast === void 0 ? void 0 : toast.bodyText
  }, /*#__PURE__*/React__default.createElement(Markdown, {
    components: {
      a: function a(_ref2) {
        var children = _ref2.children,
          href = _ref2.href;
        return /*#__PURE__*/React__default.createElement("a", {
          href: href,
          style: {
            color: linkColor,
            textDecoration: 'none'
          }
        }, children);
      },
      p: function p(_ref3) {
        var children = _ref3.children;
        return /*#__PURE__*/React__default.createElement("p", {
          style: {
            margin: 0,
            overflowWrap: 'anywhere'
          }
        }, children);
      },
      blockquote: function blockquote(_ref4) {
        var children = _ref4.children;
        return /*#__PURE__*/React__default.createElement("blockquote", {
          style: {
            margin: 0,
            paddingLeft: 10,
            borderLeft: "3px " + blockquoteColor + " solid",
            marginBottom: 5,
            marginTop: 5
          }
        }, children);
      },
      ul: function ul(_ref5) {
        var children = _ref5.children;
        return /*#__PURE__*/React__default.createElement("ul", {
          style: {
            whiteSpace: 'normal',
            margin: 0,
            paddingLeft: 10
          }
        }, children);
      },
      ol: function ol(_ref6) {
        var children = _ref6.children;
        return /*#__PURE__*/React__default.createElement("ol", {
          style: {
            whiteSpace: 'normal',
            margin: 0,
            paddingLeft: 10
          }
        }, children);
      },
      img: function img(props) {
        return /*#__PURE__*/React__default.createElement("img", _extends({
          style: {
            maxWidth: '100%',
            objectFit: 'contain'
          }
        }, props));
      }
    }
  }, message === null || message === void 0 ? void 0 : (_message$text = message.text) === null || _message$text === void 0 ? void 0 : (_message$text$replace = _message$text.replaceAll('\\\n', '&nbsp;')) === null || _message$text$replace === void 0 ? void 0 : (_message$text$replace2 = _message$text$replace.replaceAll('\n', '  \n')) === null || _message$text$replace2 === void 0 ? void 0 : _message$text$replace2.replaceAll('&nbsp;', '&nbsp;  \n')))));
}
var Container$4 = styled.div(_templateObject$7 || (_templateObject$7 = _taggedTemplateLiteralLoose(["\n  background-color: ", ";\n  cursor: pointer;\n  padding: 14px;\n  display: flex;\n  overflow-wrap: anywhere;\n"])), lightColors.main);
var HeaderText$2 = styled(CText)(_templateObject2$5 || (_templateObject2$5 = _taggedTemplateLiteralLoose(["\n  font-size: 13px;\n  line-height: 16px;\n  font-weight: 700;\n  margin: 10px 0px;\n"])));
var BodyText$1 = styled.div(_templateObject3$4 || (_templateObject3$4 = _taggedTemplateLiteralLoose(["\n  font-size: 12px;\n  line-height: 18px;\n  font-weight: 400;\n  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Helvetica,\n    Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol';\n  color: ", ";\n  margin: 10px 0px;\n  color: ", ";\n"])), lightColors.primaryText, lightColors.primaryText);
var AvatarImage$1 = styled.img(_templateObject4$3 || (_templateObject4$3 = _taggedTemplateLiteralLoose(["\n  height: 36px;\n  width: 36px;\n  border-radius: 100px;\n"])));
var AvatarView$1 = styled.div(_templateObject5$3 || (_templateObject5$3 = _taggedTemplateLiteralLoose(["\n  margin-top: 8px;\n  margin-right: 12px;\n"])));

var _templateObject$8;
function Toast(_ref) {
  var position = _ref.position,
    duration = _ref.duration,
    limit = _ref.limit;
  var toastDuration = duration || 3000;
  var toastsLimit = limit || 3;
  var toastPosition = !position ? window.innerWidth > 425 ? 'bottom-right' : 'bottom-center' : position;
  return /*#__PURE__*/React__default.createElement(StyledContainer, {
    autoClose: toastDuration,
    position: toastPosition,
    limit: toastsLimit,
    hideProgressBar: true,
    pauseOnHover: false,
    closeButton: false,
    pauseOnFocusLoss: false
  });
}
function notify(_ref2) {
  var notificationsData = _ref2.notificationsData,
    toastProps = _ref2.toastProps;
  var ToastNotificationComponent = (toastProps === null || toastProps === void 0 ? void 0 : toastProps.toastComponent) || ToastNotification;
  if (document.hidden) return;
  var toastId = reactToastify.toast( /*#__PURE__*/React__default.createElement(ToastNotificationComponent, {
    notificationData: notificationsData,
    closeToast: function closeToast() {
      reactToastify.toast.dismiss({
        id: toastId
      });
    }
  }));
}
var StyledContainer = styled(reactToastify.ToastContainer)(_templateObject$8 || (_templateObject$8 = _taggedTemplateLiteralLoose(["\n  &.Toastify__toast-container {\n    max-width: 450px;\n    min-width: 300px;\n    padding: 0px;\n    width: unset;\n  }\n\n  .Toastify__toast-theme--light,\n  .Toastify__toast-theme--default,\n  .Toastify__toast-theme--dark {\n    padding: 0px;\n    border-radius: 5px;\n    background-color: transparent;\n  }\n\n  .Toastify__toast-body {\n    padding: 0px;\n  }\n"])));

var initialNotificationData = {
  storeData: {},
  unseenCount: 0,
  hasNext: false,
  initialLoading: false,
  fetchMoreLoading: false
};
var InboxContext$1 = React.createContext({
  notificationsData: initialNotificationData
});
function SuprSendProvider(_ref) {
  var children = _ref.children,
    workspaceKey = _ref.workspaceKey,
    distinctId = _ref.distinctId,
    subscriberId = _ref.subscriberId,
    inboxId = _ref.inboxId,
    tenantId = _ref.tenantId,
    stores = _ref.stores,
    pageSize = _ref.pageSize;
  if (inboxId) {
    subscriberId = inboxId;
  }
  var _useState = React.useState(),
    inbox = _useState[0],
    setInbox = _useState[1];
  var _useState2 = React.useState(_extends({}, initialNotificationData)),
    notificationsData = _useState2[0],
    setNotificationsData = _useState2[1];
  React.useEffect(function () {
    var inboxInst = new SuprsendJSInbox(workspaceKey, {
      pageSize: pageSize,
      tenantID: tenantId,
      stores: stores
    });
    setInbox(inboxInst);
    inboxInst.emitter.on('sync_notif_store', function () {
      var inboxData = inboxInst.feed.data;
      setNotificationsData({
        storeData: (inboxData === null || inboxData === void 0 ? void 0 : inboxData.stores) || {},
        count: (inboxData === null || inboxData === void 0 ? void 0 : inboxData.unseenCount) || 0,
        hasNext: inboxData === null || inboxData === void 0 ? void 0 : inboxData.hasNext,
        initialLoading: inboxData === null || inboxData === void 0 ? void 0 : inboxData.initialLoading,
        fetchMoreLoading: inboxData === null || inboxData === void 0 ? void 0 : inboxData.fetchMoreLoading,
        activeStoreId: inboxData === null || inboxData === void 0 ? void 0 : inboxData.activeStoreId
      });
    });
    return function () {
      inboxInst.emitter.all.clear();
    };
  }, []);
  React.useEffect(function () {
    if (!subscriberId || !inbox) return;
    inbox.identifyUser(distinctId, subscriberId);
    inbox.feed.fetchNotifications();
    return function () {
      inbox.resetUser();
    };
  }, [subscriberId, inbox]);
  return /*#__PURE__*/React__default.createElement(InboxContext$1.Provider, {
    value: {
      inbox: inbox,
      notificationsData: notificationsData,
      subscriberId: subscriberId
    }
  }, children);
}

function useNotifications(storeId) {
  var _inboxData$inbox, _notificationsFeed$st, _notificationsFeed$st2, _inboxData$notificati, _inboxData$notificati2;
  var inboxData = React.useContext(InboxContext$1);
  var notificationsFeed = (_inboxData$inbox = inboxData.inbox) === null || _inboxData$inbox === void 0 ? void 0 : _inboxData$inbox.feed;
  React.useEffect(function () {
    if (inboxData !== null && inboxData !== void 0 && inboxData.inbox && storeId) {
      inboxData.inbox.changeActiveStore(storeId);
    }
  }, [storeId, inboxData === null || inboxData === void 0 ? void 0 : inboxData.inbox, inboxData === null || inboxData === void 0 ? void 0 : inboxData.subscriberId]);
  var firstStore = (notificationsFeed === null || notificationsFeed === void 0 ? void 0 : (_notificationsFeed$st = notificationsFeed.stores) === null || _notificationsFeed$st === void 0 ? void 0 : (_notificationsFeed$st2 = _notificationsFeed$st[0]) === null || _notificationsFeed$st2 === void 0 ? void 0 : _notificationsFeed$st2.storeId) || 'default_store';
  var activeStoreId = storeId || firstStore;
  var activeStoreData = inboxData === null || inboxData === void 0 ? void 0 : (_inboxData$notificati = inboxData.notificationsData) === null || _inboxData$notificati === void 0 ? void 0 : (_inboxData$notificati2 = _inboxData$notificati.storeData) === null || _inboxData$notificati2 === void 0 ? void 0 : _inboxData$notificati2[activeStoreId];
  return {
    notifications: (activeStoreData === null || activeStoreData === void 0 ? void 0 : activeStoreData.notifications) || [],
    initialLoading: inboxData.notificationsData.initialLoading,
    hasNext: inboxData.notificationsData.hasNext,
    fetchMoreLoading: inboxData.notificationsData.fetchMoreLoading,
    markClicked: notificationsFeed === null || notificationsFeed === void 0 ? void 0 : notificationsFeed.markClicked.bind(notificationsFeed),
    markRead: notificationsFeed === null || notificationsFeed === void 0 ? void 0 : notificationsFeed.markRead.bind(notificationsFeed),
    markUnRead: notificationsFeed === null || notificationsFeed === void 0 ? void 0 : notificationsFeed.markUnRead.bind(notificationsFeed),
    markArchived: notificationsFeed === null || notificationsFeed === void 0 ? void 0 : notificationsFeed.markArchived.bind(notificationsFeed),
    markAllSeen: notificationsFeed === null || notificationsFeed === void 0 ? void 0 : notificationsFeed.markAllSeen.bind(notificationsFeed),
    markAllRead: notificationsFeed === null || notificationsFeed === void 0 ? void 0 : notificationsFeed.markAllRead.bind(notificationsFeed),
    fetchPrevious: notificationsFeed === null || notificationsFeed === void 0 ? void 0 : notificationsFeed.fetchNotifications.bind(notificationsFeed)
  };
}

function useBell() {
  var _inboxData$inbox, _inboxData$inbox$feed, _inboxData$inbox2;
  var inboxData = React.useContext(InboxContext$1);
  return {
    unSeenCount: inboxData.notificationsData.count,
    markAllSeen: (_inboxData$inbox = inboxData.inbox) === null || _inboxData$inbox === void 0 ? void 0 : (_inboxData$inbox$feed = _inboxData$inbox.feed.markAllSeen) === null || _inboxData$inbox$feed === void 0 ? void 0 : _inboxData$inbox$feed.bind((_inboxData$inbox2 = inboxData.inbox) === null || _inboxData$inbox2 === void 0 ? void 0 : _inboxData$inbox2.feed)
  };
}
function useUnseenCount() {
  var _inboxData$inbox3, _inboxData$inbox3$fee, _inboxData$inbox4;
  var inboxData = React.useContext(InboxContext$1);
  return {
    unSeenCount: inboxData.notificationsData.count,
    markAllSeen: (_inboxData$inbox3 = inboxData.inbox) === null || _inboxData$inbox3 === void 0 ? void 0 : (_inboxData$inbox3$fee = _inboxData$inbox3.feed.markAllSeen) === null || _inboxData$inbox3$fee === void 0 ? void 0 : _inboxData$inbox3$fee.bind((_inboxData$inbox4 = inboxData.inbox) === null || _inboxData$inbox4 === void 0 ? void 0 : _inboxData$inbox4.feed)
  };
}

function useEvent(eventName, callback) {
  var inboxData = React.useContext(InboxContext$1);
  React.useEffect(function () {
    var _inboxData$inbox;
    var emitter = inboxData === null || inboxData === void 0 ? void 0 : (_inboxData$inbox = inboxData.inbox) === null || _inboxData$inbox === void 0 ? void 0 : _inboxData$inbox.emitter;
    if (emitter) {
      emitter.on(eventName, callback);
    }
    return function () {
      return emitter === null || emitter === void 0 ? void 0 : emitter.off(eventName);
    };
  }, [inboxData === null || inboxData === void 0 ? void 0 : inboxData.inbox]);
}

function useStoreUnseenCount(storeId) {
  var _inboxData$notificati, _inboxData$notificati2;
  var inboxData = React.useContext(InboxContext$1);
  return {
    unSeenCount: (_inboxData$notificati = inboxData.notificationsData.storeData) === null || _inboxData$notificati === void 0 ? void 0 : (_inboxData$notificati2 = _inboxData$notificati[storeId]) === null || _inboxData$notificati2 === void 0 ? void 0 : _inboxData$notificati2.unseenCount
  };
}
function useStoresUnseenCount() {
  var inboxData = React.useContext(InboxContext$1);
  var storeData = inboxData.notificationsData.storeData;
  var storeCount = {};
  for (var storeId in storeData) {
    var store = storeData[storeId];
    storeCount[storeId] = store.unseenCount;
  }
  return storeCount;
}

function SuprsendInbox(_ref) {
  var _ref$workspaceKey = _ref.workspaceKey,
    workspaceKey = _ref$workspaceKey === void 0 ? '' : _ref$workspaceKey,
    _ref$tenantId = _ref.tenantId,
    tenantId = _ref$tenantId === void 0 ? '' : _ref$tenantId,
    _ref$distinctId = _ref.distinctId,
    distinctId = _ref$distinctId === void 0 ? '' : _ref$distinctId,
    _ref$subscriberId = _ref.subscriberId,
    subscriberId = _ref$subscriberId === void 0 ? '' : _ref$subscriberId,
    _ref$inboxId = _ref.inboxId,
    inboxId = _ref$inboxId === void 0 ? '' : _ref$inboxId,
    stores = _ref.stores,
    toastProps = _ref.toastProps,
    notificationClickHandler = _ref.notificationClickHandler,
    bellComponent = _ref.bellComponent,
    badgeComponent = _ref.badgeComponent,
    notificationComponent = _ref.notificationComponent,
    noNotificationsComponent = _ref.noNotificationsComponent,
    loaderComponent = _ref.loaderComponent,
    tabBadgeComponent = _ref.tabBadgeComponent,
    _ref$hideAvatar = _ref.hideAvatar,
    hideAvatar = _ref$hideAvatar === void 0 ? false : _ref$hideAvatar,
    hideInbox = _ref.hideInbox,
    hideToast = _ref.hideToast,
    theme = _ref.theme,
    _ref$themeType = _ref.themeType,
    themeType = _ref$themeType === void 0 ? 'light' : _ref$themeType,
    _ref$pagination = _ref.pagination,
    pagination = _ref$pagination === void 0 ? true : _ref$pagination,
    pageSize = _ref.pageSize,
    _ref$popperPosition = _ref.popperPosition,
    popperPosition = _ref$popperPosition === void 0 ? 'bottom' : _ref$popperPosition,
    _ref$showUnreadCountO = _ref.showUnreadCountOnTabs,
    showUnreadCountOnTabs = _ref$showUnreadCountO === void 0 ? true : _ref$showUnreadCountO;
  if (inboxId) {
    subscriberId = inboxId;
  }
  var _useState = React.useState(false),
    openInbox = _useState[0],
    toggleInbox = _useState[1];
  var _useState2 = React.useState(),
    inbox = _useState2[0],
    setInbox = _useState2[1];
  var _useState3 = React.useState({
      storeData: {},
      count: 0
    }),
    notificationsData = _useState3[0],
    setNotificationsData = _useState3[1];
  var _useState4 = React.useState(''),
    activeStore = _useState4[0],
    setActiveStore = _useState4[1];
  var _useState5 = React.useState(false),
    changingActiveStore = _useState5[0],
    setChangingActiveStore = _useState5[1];
  React.useEffect(function () {
    var inboxInst = new SuprsendJSInbox(workspaceKey, {
      pageSize: pageSize,
      tenantID: tenantId,
      stores: stores
    });
    setInbox(inboxInst);
    inboxInst.emitter.on('sync_notif_store', function () {
      var inboxData = inboxInst.feed.data;
      setNotificationsData({
        storeData: (inboxData === null || inboxData === void 0 ? void 0 : inboxData.stores) || {},
        count: (inboxData === null || inboxData === void 0 ? void 0 : inboxData.unseenCount) || 0,
        hasNext: inboxData === null || inboxData === void 0 ? void 0 : inboxData.hasNext,
        initialLoading: inboxData === null || inboxData === void 0 ? void 0 : inboxData.initialLoading,
        activeStoreId: inboxData === null || inboxData === void 0 ? void 0 : inboxData.activeStoreId
      });
      setActiveStore(inboxData === null || inboxData === void 0 ? void 0 : inboxData.activeStoreId);
    });
    inboxInst.emitter.on('new_notification', function (notification) {
      notify({
        notificationsData: notification,
        toastProps: toastProps
      });
    });
    return function () {
      inboxInst.emitter.all.clear();
    };
  }, []);
  React.useEffect(function () {
    if (openInbox && subscriberId && inbox) {
      inbox.feed.markAllSeen();
    }
  }, [openInbox, subscriberId, inbox]);
  React.useEffect(function () {
    if (!subscriberId || !inbox) return;
    setNotificationsData({
      storeData: {},
      count: 0
    });
    inbox.identifyUser(distinctId, subscriberId);
    inbox.feed.fetchNotifications();
    return function () {
      inbox.resetUser();
    };
  }, [subscriberId, inbox]);
  var themeValue = themeType === 'dark' ? mergeDeep(darkTheme, theme) : theme || {};
  return /*#__PURE__*/React__default.createElement(InboxThemeContext.Provider, {
    value: themeValue
  }, /*#__PURE__*/React__default.createElement(InboxContext.Provider, {
    value: {
      workspaceKey: workspaceKey,
      notificationsData: notificationsData,
      setNotificationsData: setNotificationsData,
      notificationClickHandler: notificationClickHandler,
      bellComponent: bellComponent,
      badgeComponent: badgeComponent,
      notificationComponent: notificationComponent,
      noNotificationsComponent: noNotificationsComponent,
      tabBadgeComponent: tabBadgeComponent,
      toggleInbox: toggleInbox,
      inbox: inbox,
      loaderComponent: loaderComponent,
      pagination: pagination,
      hideAvatar: hideAvatar,
      activeStore: activeStore,
      setActiveStore: setActiveStore,
      changingActiveStore: changingActiveStore,
      setChangingActiveStore: setChangingActiveStore,
      showUnreadCountOnTabs: showUnreadCountOnTabs,
      themeType: themeType
    }
  }, !hideInbox && /*#__PURE__*/React__default.createElement(Inbox, {
    openInbox: openInbox,
    toggleInbox: toggleInbox,
    popperPosition: popperPosition
  }), !hideToast && /*#__PURE__*/React__default.createElement(Toast, toastProps)));
}

exports.SuprSendProvider = SuprSendProvider;
exports.default = SuprsendInbox;
exports.useBell = useBell;
exports.useEvent = useEvent;
exports.useNotifications = useNotifications;
exports.useStoreUnseenCount = useStoreUnseenCount;
exports.useStoresUnseenCount = useStoresUnseenCount;
exports.useUnseenCount = useUnseenCount;
